### PR TITLE
APPS-1248 Add estimate time to restore

### PR DESCRIPTION
**Pull request by korotkov-aerospike**

### CODE REVIEW

File: pkg/model/current_backup.go [hunk 1]
```diff
@@ -2,73 +2,28 @@ package model
...
+	// TotalRecords: the total number of records to be processed.
 	TotalRecords uint64 `json:"total-records,omitempty" example:"100"`
-	// DoneRecords: the number of records that have been successfully backed up.
+	// DoneRecords: the number of records that have been successfully done.
 	DoneRecords uint64 `json:"done-records,omitempty" example:"50"`
 	// StartTime: the time when the backup operation started.
 	StartTime time.Time `json:"start-time,omitempty" example:"2006-01-02T15:04:05Z07:00"`
 	// PercentageDone: the progress of the backup operation as a percentage.
-	PercentageDone int `json:"percentage-done,omitempty" example:"50"`
+	PercentageDone uint `json:"percentage-done,omitempty" example:"50"`
```
Comment (reugn) [inline pkg/model/current_backup.go:25]: Don't we want to support percent fractions (two digits) using a float32?

File: pkg/model/restore_policy.go [hunk 2]
```diff
@@ -46,12 +46,9 @@ type RestorePolicy struct {
 	// With this option, records from the backup always overwrite records that already exist in
 	// the namespace, regardless of generation numbers.
 	NoGeneration *bool `json:"no-generation,omitempty"`
-	// Throttles read operations from the backup file(s) to not exceed the given I/O bandwidth
-	// in MiB/s and its database write operations to not exceed the given number of transactions
-	// per second.
+	// Throttles read operations from the backup file(s) to not exceed the given I/O bandwidth in bytes/sec
```
Comment (reugn) [inline pkg/model/restore_policy.go:49]: Add periods to be consistent.

File: pkg/model/restore_result.go [hunk 3]
```diff
@@ -11,14 +11,14 @@ const (
...
+	RestoreStats
+	CurrentRestore *RunningJob `yaml:"current-restore,omitempty" json:"current-job,omitempty"`
+	Status         JobStatus   `yaml:"status,omitempty" json:"status,omitempty" enums:"Running,Done,Failed"`
+	Error          string      `yaml:"error,omitempty" json:"error,omitempty"`
 }
 
-// RestoreResult represents a single restore operation result.
-type RestoreResult struct {
-	TotalRecords    uint64 `yaml:"total-records,omitempty" json:"total-records,omitempty" format:"int64" example:"10"`
+type RestoreStats struct {
```
Comment (reugn) [inline pkg/model/restore_result.go:20]: Add a doc.

File: pkg/service/estimates.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+func getAnyHandler(m map[string]*backup.BackupHandler) *backup.BackupHandler {
+	for _, value := range m {
+		return value
+	}
+
+	return nil
+}
+
+func RestoreJobStatus(job *jobInfo) *model.RestoreJobStatus {
```
Comment (reugn) [inline pkg/service/estimates.go:35]: Add a doc, explaining when and for which purpose the function is used.

File: pkg/service/restore_data.go [hunk 5]
```diff
@@ -58,12 +60,21 @@ func (r *dataRestorer) Restore(request *model.RestoreRequestInternal) (RestoreJo
...
+		handler, err := r.restoreService.RestoreRun(ctx, client, request)
+		if err != nil {
+			r.restoreJobs.setFailed(jobID, fmt.Errorf("failed to start restore operation: %w", err))
+			return
+		}
+		r.restoreJobs.addTotalRecords(jobID, totalRecords)
+		r.restoreJobs.addHandler(jobID, handler)
+
+		// Wait for the restore operation to complete
+		err = handler.Wait(ctx)
```
Comment (reugn) [inline pkg/service/restore_data.go:72]: Update to the latest version of the library.

File: pkg/service/s3_context.go [hunk 6]
```diff
@@ -292,18 +292,23 @@ func (s *S3Context) wrapWithPrefix(path string) *string {
...
-func (s *S3Context) validateStorageContainsBackup() error {
+func (s *S3Context) ValidateStorageContainsBackup() (uint64, error) {
 	files, err := s.lsFiles(s.path)
 	if err != nil {
-		return err
+		return 0, err
 	}
 	if len(files) == 0 {
-		return fmt.Errorf("given path %s not exist", s.path)
+		return 0, fmt.Errorf("given path %s not exist", s.path)
```
Comment (reugn) [inline pkg/service/s3_context.go:301]: .. does not

File: pkg/model/current_backup.go [hunk 7]
```diff
@@ -2,73 +2,28 @@ package model
...
 
-type CurrentBackup struct {
-	// TotalRecords: the total number of records to be backed up.
+// RunningJob tracks progress of currently running job.
+// @Description RunningJob tracks progress of currently running job.
+type RunningJob struct {
+	// TotalRecords: the total number of records to be processed.
 	TotalRecords uint64 `json:"total-records,omitempty" example:"100"`
-	// DoneRecords: the number of records that have been successfully backed up.
+	// DoneRecords: the number of records that have been successfully done.
```
[the remark this answers is not in the document]
Comment (korotkov-aerospike) [reply pkg/model/current_backup.go:20]: Backed up or restored. We use same structure for both.

File: pkg/model/current_backup.go (hunk 1 above)
Comment (korotkov-aerospike) [reply pkg/model/current_backup.go:25]: I think, integer percent is enough, why bother with floats.

File: pkg/service/restore_data.go (hunk 5 above)
Comment (korotkov-aerospike) [reply pkg/service/restore_data.go]: wow, this update did helped fix the unit test, but caused a lot of diffs.

**Review by reugn — APPROVED** [review]

File: pkg/model/restore_result.go [hunk 8]
```diff
@@ -11,14 +11,15 @@ const (
 // RestoreJobStatus represents a restore job status.
 // @Description RestoreJobStatus represents a restore job status.
 type RestoreJobStatus struct {
-	RestoreResult
-	Status JobStatus `yaml:"status,omitempty" json:"status,omitempty" enums:"Running,Done,Failed"`
-	Error  string    `yaml:"error,omitempty" json:"error,omitempty"`
+	RestoreStats
+	CurrentRestore *RunningJob `yaml:"current-restore,omitempty" json:"current-job,omitempty"`
+	Status         JobStatus   `yaml:"status,omitempty" json:"status,omitempty" enums:"Running,Done,Failed"`
+	Error          string      `yaml:"error,omitempty" json:"error,omitempty"`
```
Comment (filkeith) [inline pkg/model/restore_result.go:17]: it is string because we return it as rest response? Maybe it should be ErrorMessage

**Review by filkeith — APPROVED** [review]
lgtm in total

**Pull request merged**